compiler-rt/lib/profile/InstrProfilingWriter.c

```diff
@@ -273,18 +280,28 @@ lprofWriteDataImpl(ProfDataWriter *Writer, const __llvm_profile_data *DataBegin,
   const uint64_t NumBitmapBytes =
       __llvm_profile_get_num_bitmap_bytes(BitmapBegin, BitmapEnd);
   const uint64_t NamesSize = __llvm_profile_get_name_size(NamesBegin, NamesEnd);
+  const uint64_t NumVTables =
+      __llvm_profile_get_num_vtable(VTableBegin, VTableEnd);
+  const uint64_t VTableSectionSize =
+      __llvm_profile_get_vtable_section_size(VTableBegin, VTableEnd);
+  const uint64_t VNamesSize =
+      __llvm_profile_get_name_size(VNamesBegin, VNamesEnd);
 
   /* Create the header. */
   __llvm_profile_header Header;
 
   /* Determine how much padding is needed before/after the counters and after
    * the names. */
   uint64_t PaddingBytesBeforeCounters, PaddingBytesAfterCounters,
-      PaddingBytesAfterNames, PaddingBytesAfterBitmapBytes;
-  __llvm_profile_get_padding_sizes_for_counters(
-      DataSectionSize, CountersSectionSize, NumBitmapBytes, NamesSize,
-      &PaddingBytesBeforeCounters, &PaddingBytesAfterCounters,
-      &PaddingBytesAfterBitmapBytes, &PaddingBytesAfterNames);
+      PaddingBytesAfterBitmapBytes, PaddingBytesAfterNames,
+      PaddingBytesAfterVTable, PaddingBytesAfterVNames;
+  if (__llvm_profile_get_padding_sizes_for_counters(
```

Comment:
`__llvm_profile_get_padding_sizes_for_counters` return both `0` and `-1` on failure and this here is only checking for -1. Is that intentional?

If the tri-state is needed I would be very explicit on having an enum for `success`, `failure`, or `error`. Otherwise, would 2 states be enough?

Reply:
`__llvm_profile_get_padding_sizes_for_counters` returns -1 to alert the [incompatibility](https://github.com/llvm/llvm-project/pull/81691/files/0c9ed57bce7d29f27ab7562d85503edfa8132670#diff-ff206eae21de963bf60f9baca54d2c90e0dd89bf53732c236b019b4e9192bd20R163) and returns 0 otherwise.

-1 is chosen to signal errors to for consistency with existing code (e.g.,  [`InstrProfilingWriter.c`](https://github.com/llvm/llvm-project/blob/118a2a52fd465b7576c54bd102ee2e417a3b9a71/compiler-rt/lib/profile/InstrProfilingWriter.c) uses -1 to return error in many places, and [InstrProfilingFile.c:writeFile](https://github.com/llvm/llvm-project/blob/118a2a52fd465b7576c54bd102ee2e417a3b9a71/compiler-rt/lib/profile/InstrProfilingFile.c#L518) also returns -1). Non-zero value is handled [here](https://github.com/llvm/llvm-project/blob/118a2a52fd465b7576c54bd102ee2e417a3b9a71/compiler-rt/lib/profile/InstrProfilingFile.c#L605-L606) to emit warnings.

Anecdotally, `llvm_unreachable` is not a way to signal errors in release build with the default LLVM config.
 * https://llvm.org/docs/CodingStandards.html#assert-liberally mentions this _When assertions are disabled (i.e. in release builds), llvm_unreachable becomes a hint to compilers to skip generating code for this branch_ 
 * https://llvm.org/docs/CMake.html#llvm-related-variables mentions `DLLVM_UNREACHABLE_OPTIMIZE` is on by default and enables the hint

Reply:
Also, compiler-rt/lib/profile is in C (but not C++ yet afaik, see https://discourse.llvm.org/t/using-c-and-sanitizer-common-in-profile-runtime/58629) so we don't have `bool` type either.

Reply:
I see, even though there's precedent for using -1 I'm personally not a big fan because it's an uncommon pattern. Also I've usually found for many llvm functions that a `bool` of `true` indicates success and `false` failure which causes the opposite pattern (an explicit `bool` type isn't necessary for the same behavior).

If `unreachable` was considered as a way to signal error can error out immediately at the fail site? Naively a direct termination, if there's an error handle around invoking that would be good.

Reply:
To clarify a little bit, we don't want to use `llvm_unreachable` to signal errors at least for release builds.

> Also I've usually found for many llvm functions that a bool of true indicates success and false failure which causes the opposite pattern (an explicit bool type isn't necessary for the same behavior).

I agree that using enums has better readability in general. However, given the callers on the chain of the current function regards non-zero (concretely -1) as error and 0 as ok, it's a larger change (that usually goes to a separate PR) to improve the readability using enums. The `true` and `false` interpretation in llvm depends a lot on the caller context based on my experience.